### PR TITLE
Refactor stablizer

### PR DESCRIPTION
**Pull request by monicsh**

### CODE REVIEW

File: KVStoreAlgorithm.cpp [hunk 1]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
- * FUNCTION NAME: stabilizationProtocol
- *
- * DESCRIPTION: This runs the stabilization protocol in case of Node joins and leaves
- *                              It ensures that there always 3 copies of all keys in the DHT at all times
- *                              The function does the following:
- *                              1) Ensures that there are three "CORRECT" replicas of all the keys in spite of failures and joins
- *                              Note:- "CORRECT" replicas implies that every key is replicated in its two neighboring nodes in the ring
+ * Required fields for Stablizer
+ m_ring
+m_memberNode
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: add * for comment

File: KVStoreAlgorithm.cpp [hunk 2]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
-    size_t succ_2 = (i+2)%(this->m_ring.size());
-    size_t pred_1 = (i-1 + this->m_ring.size())%(this->m_ring.size());
-    size_t pred_2 = (i-2 + this->m_ring.size())%(this->m_ring.size());
+size_t KVStoreAlgorithm::findfirstSuccessorIndex(size_t myPos){
+    return (myPos+1)%(this->m_ring.size());
+}
 
-    // Initialize successors and predeccesors
-    if (this->m_hasMyReplicas.empty() && this->m_haveReplicasOf.empty()){
+size_t KVStoreAlgorithm::findSecondSuccessorIndex(size_t myPos){
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: do this way
if (myPos == -1) 
{
return -1;
}

File: KVStoreAlgorithm.cpp [hunk 3]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+    return myPos;
+}
 
-    // set successors and predeccesors index
-    size_t succ_1 = (i+1)%(this->m_ring.size());
-    size_t succ_2 = (i+2)%(this->m_ring.size());
-    size_t pred_1 = (i-1 + this->m_ring.size())%(this->m_ring.size());
-    size_t pred_2 = (i-2 + this->m_ring.size())%(this->m_ring.size());
+size_t KVStoreAlgorithm::findfirstSuccessorIndex(size_t myPos){
+    return (myPos+1)%(this->m_ring.size());
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: do this way
if (myPos == -1)
{
return -1;
}

File: KVStoreAlgorithm.cpp (hunk 2 above)
Comment (kgaipal) [reply KVStoreAlgorithm.cpp]: infact find what is max of size_t and tturn that check in std::size_t::max or tsomething

File: KVStoreAlgorithm.cpp [hunk 4]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+}
 
-    // Initialize successors and predeccesors
-    if (this->m_hasMyReplicas.empty() && this->m_haveReplicasOf.empty()){
+size_t KVStoreAlgorithm::findSecondSuccessorIndex(size_t myPos){
+    return (myPos+2)%(this->m_ring.size());
+}
 
+size_t KVStoreAlgorithm::findfirstPredeccesorIndex(size_t myPos){
+    return (myPos-1 + this->m_ring.size())%(this->m_ring.size());
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: same check

File: KVStoreAlgorithm.cpp [hunk 5]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+}
+
+size_t KVStoreAlgorithm::findSecondPredeccesorIndex(size_t myPos){
+    return (myPos-2 + this->m_ring.size())%(this->m_ring.size());
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: what is myPos is less than 2?

File: KVStoreAlgorithm.cpp [hunk 6]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+
+void KVStoreAlgorithm::setHaveReplicasOf(size_t pred_1, size_t pred_2){
+    // Initialize successors and predeccesors
+    if (this->m_haveReplicasOf.empty()){
         this->m_haveReplicasOf.push_back(m_ring[pred_1]);
         this->m_haveReplicasOf.push_back(m_ring[pred_2]);
     }
+}
+
+void KVStoreAlgorithm::remakeReplicaSetImPrimary(
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: typoe 

remakeReplicaSetInPrimary

File: KVStoreAlgorithm.cpp [hunk 7]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+}
+
+size_t KVStoreAlgorithm::findSecondPredeccesorIndex(size_t myPos){
+    return (myPos-2 + this->m_ring.size())%(this->m_ring.size());
+}
+
+void KVStoreAlgorithm::setHasMyReplicas(size_t succ_1, size_t succ_2){
+    // Initialize successors
+    if (this->m_hasMyReplicas.empty()){
         this->m_hasMyReplicas.push_back(m_ring[succ_1]);
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: if (! this->m_hasMyReplicas.empty()){
return;
}
       
this->m_hasMyReplicas.push_back(m_ring[succ_1]);
this->m_hasMyReplicas.push_back(m_ring[succ_2]);

Comment (kgaipal) [reply KVStoreAlgorithm.cpp]: always do defensive coding

File: KVStoreAlgorithm.cpp [hunk 8]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
 
+size_t KVStoreAlgorithm::findfirstPredeccesorIndex(size_t myPos){
+    return (myPos-1 + this->m_ring.size())%(this->m_ring.size());
+}
+
+size_t KVStoreAlgorithm::findSecondPredeccesorIndex(size_t myPos){
+    return (myPos-2 + this->m_ring.size())%(this->m_ring.size());
+}
+
+void KVStoreAlgorithm::setHasMyReplicas(size_t succ_1, size_t succ_2){
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: rename:

saveReplicasIfEmpty() ... something likfe that

File: KVStoreAlgorithm.cpp [hunk 9]
```diff
@@ -653,46 +653,122 @@ bool KVStoreAlgorithm::recvLoop()
...
+ *                              It ensures that there always 3 copies of all keys in the DHT at all times
+ *                              The function does the following:
+ *                              1) Ensures that there are three "CORRECT" replicas of all the keys in spite of failures and joins
+ *                              Note:- "CORRECT" replicas implies that every key is replicated in its two neighboring nodes in the ring
+ */
+void KVStoreAlgorithm::stabilizationProtocol()
+{
+    size_t myPositionInRing = myPositionInTheRing();
+
+    if (myPositionInRing == -1)
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: < 0

File: KVStoreAlgorithm.cpp (hunk 1 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: removed comments

File: KVStoreAlgorithm.cpp (hunk 3 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: done

File: KVStoreAlgorithm.cpp (hunk 2 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: convert datatype from size_t to int; size_t is unsigned datatype, so not using for position variable.

File: KVStoreAlgorithm.cpp (hunk 4 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: changed size_t to int And add statemnet for verifing negative myPos value.

File: KVStoreAlgorithm.cpp (hunk 5 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: < 0 value teruen as -1

File: KVStoreAlgorithm.cpp (hunk 6 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: renamed!

File: KVStoreAlgorithm.cpp (hunk 9 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: done

File: KVStoreAlgorithm.cpp (hunk 7 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: Done!

File: KVStoreAlgorithm.cpp (hunk 8 above)
Comment (monicsh) [reply KVStoreAlgorithm.cpp]: Done

File: KVStoreAlgorithm.cpp [hunk 10]
```diff
@@ -652,138 +652,194 @@ bool KVStoreAlgorithm::recvLoop()
...
-{
+int KVStoreAlgorithm::myPositionInTheRing(){
     // 1. find out my postition in the ring
-    size_t myPos = -1;
-    int i;
-    for (i = 0; i < this->m_ring.size(); i++) {
+    int myPos = -1;
+    for (int i = 0; i < this->m_ring.size(); i++) {
         if (m_ring[i].nodeAddress == this->m_memberNode->addr){
             myPos = i;          // found pos
```
Comment (kgaipal) [inline KVStoreAlgorithm.cpp]: return myPos;

Comment (kgaipal) [reply KVStoreAlgorithm.cpp]: no need to berak

**Comment by kgaipal** [comment]
Ok

**Pull request merged**